[TaskGroup] Reenable test and fix memory issue

ktoso opened the pull request:
**Description:** This re-enables a task group test and fixes a memory issue where we realized we must release the group lock before resuming the waiting task. If we don't do this, we're racing the group's destroy() (coming from the parent task, waiting on waitAll with the unlock()), risking calling unlock() on already released memory - very rarely causing crashes due to this. It is hard to reproduce this issue, due to the race appearing very rarely, but it is consistently reproducible in some environments.
**Risk:** Low, the fix reorders an unlock() with the waitingTask resume which cannot cause any other consistency issues as there are no other actions taken after the resume. Arguably, this always should have been release + run order.
**Reward:** Medium, it seems this crash does not manifest frequently in real applications, however it is a real issue we need to fix.
**Review by:** @al45tair @erik
**Testing:** CI testing, confirmed the fix on specially prepared environments.
**Radar:** rdar://112759360 rdar://110474311

ktoso committed:
Test reenable: async_taskgroup_discarding_dontLeak_class_error.swift

ktoso commented:
@swift-ci please smoke test

al45tair commented on test/Concurrency/Runtime/async_taskgroup_discarding_dontLeak_class_error.swift:
```diff
@@ -12,35 +12,75 @@
+    defer { count -= 1 }
+    if count == 0 {
+      fatalError("Counted down more times than expected! (From: \(from))")
+    } else if count > 0 {
```
This should probably be `count == 1` rather than `count > 0`.

ktoso replied:
Thank you, fixed

ktoso commented on test/Concurrency/Runtime/async_taskgroup_discarding_dontLeak_class_error.swift:
```diff
@@ -12,35 +12,75 @@
+    defer { count -= 1 }
+    if count == 0 {
+      fatalError("Counted down more times than expected! (From: \(from))")
+    } else if count > 0 {
```
```suggestion
    } else if count == 1 {
```

Whoops thank you

ktoso commented:
@swift-ci please smoke test

etcwilde commented on test/Concurrency/Runtime/async_taskgroup_discarding_dontLeak_class_error.swift:
```diff
@@ -12,35 +12,75 @@
+    print("done") // CHECK: done
   }
-}
+}
```
missing end-of-file newline here. :p

ktoso replied:
Thanks I'll add, seems we have repeatedly crashing linux on CI here... it's pretty weird as cannot reproduce on other Linux instances.

ktoso replied:
```suggestion
}

```

ktoso commented:
```
41849 --
41850 Exit Code: 2
41851
41852 Command Output (stdout):
41853 --
41854 /home/build-user/build/buildbot_linux/swift-linux-x86_64/test-linux-x86_64/Concurrency/Runtime/Output/async_taskgroup_discarding_dontLeak_class_error.swift.tmp/a.out
41855
41856 --
41857 Command Output (stderr):
41858 --
41859 /home/build-user/swift/test/Concurrency/Runtime/async_taskgroup_discarding_dontLeak_class_error.swift:73:9: warning: integer literal is unused
41860         12 // ignore this on purpose
41861         ^~
41862 Swift/ContiguousArrayBuffer.swift:613: Fatal error: Index out of range
41863 Current stack trace:
41864 0    libswiftCore.so                    0x00007fd4e0588950 _swift_stdlib_reportFatalErrorInFile + 113
41865 1    libswiftCore.so                    0x00007fd4e0268114 <unavailable> + 1458452
41866 2    libswiftCore.so                    0x00007fd4e0267f94 <unavailable> + 1458068
41867 3    libswiftCore.so                    0x00007fd4e0267cca <unavailable> + 1457354
41868 4    libswiftCore.so                    0x00007fd4e0267860 _assertionFailure(_:_:file:line:flags:) + 253
41869 5    libswift_Backtracing.so            0x00007fd4e084962e <unavailable> + 558638
41870 6    libswift_Backtracing.so            0x00007fd4e07effd0 static Backtrace.capture<A, B>(from:using:images:algorithm:limit:offset:top:) + 1915
41871 7                                       0x000055e1ed5bb0f7 <unavailable> + 82167
41872 8                                       0x000055e1ed5ba96b <unavailable> + 80235
41873 9                                       0x000055e1ed5b0b16 <unavailable> + 39702
41874 10                                      0x000055e1ed5b9c76 <unavailable> + 76918
41875 11   libc.so.6                          0x00007fd4dfbe9f90 __libc_start_main + 243
41876 12                                      0x000055e1ed5afc6e <unavailable> + 35950
41877 FileCheck error: '<stdin>' is empty.
41878 FileCheck command line:  /home/build-user/build/buildbot_linux/llvm-linux-x86_64/bin/FileCheck --allow-unused-prefixes /home/build-user/swift/test/Concurrency/Runtime/async_taskgroup_discarding_dontLeak_class_error.swift --dump-input=always                                                                                                                                                                         41879
41880 --
```

storing the crash here for reference

ktoso commented:
@swift-ci please smoke test linux

ktoso commented:
@swift-ci please smoke test linux

ktoso committed:
Unlock BEFORE resuming the waitAll waiting task, to avoid unlock() on
released memory.

Because the waitAll() group method is called before the group returns
from withTaskGroup and is used to ensure all tasks have completed before
we destroy the task group, this method is then immediately followed by
calling TaskGroup::destroy.

If we, as previously, first resume the waiting task and then attempt to
unlock the lock held by the group, we are in an unsafe situation with
racing the task group destroy() and the unlock().

Therefore, we must release the lock before we resume the waiting task.

ktoso commented:
@swift-ci please test

tomerd commented on include/swift/AST/ASTContext.h:
```diff
@@ -898,6 +898,10 @@ class ASTContext final {
 
+  /// Get the runtime availability of the `DiscardingTaskGroup`,
+  /// and supporting runtime functions function
+  AvailabilityContext getConcurrencyDiscardingTaskGroupAvailability();
```
is this related to the fix?

ktoso replied:
No, thank you -- rebase should get rid of this. Cleaning that up now

tshortli commented:
> Confirmed that this was only a flaky test.

Is the PR description out-of-date? Isn't an important correctness fix in the runtime included?

ktoso commented:
> Is the PR description out-of-date? Isn't an important correctness fix in the runtime included?

Yes, updated to CCC style with all the details from https://github.com/apple/swift/pull/67700 now

ktoso commented:
@swift-ci please smoke test

ktoso commented:
Rebased, this has only the necessary change right now.

ktoso commented:
@swift-ci please smoke test Windows

ktoso commented:
@swift-ci please smoke test Linux 

(CI/github connection issues)

ktoso commented:
@swift-ci please smoke test Linux

ktoso commented:
@swift-ci please smoke test Linux

ktoso commented:
@swift-ci please smoke test macOS

ktoso commented:
Failed Tests (1):
  Swift(macosx-x86_64) :: stdlib/Observation/ObservableAvailabilityCycle.swift

ktoso commented:
@swift-ci please smoke test macOS and merge

ktoso commented:
@swift-ci please smoke test and merge macOS

ktoso commented:
@swift-ci please smoke test macOS
